### ISSUE TITLE
topKValues and topKIndices for GPU

### ISSUE THREAD

**Issue by nsthorat** [tensorflow/tfjs#94]
_From @chaosmail on December 29, 2017 20:25_

Placeholder issue for `topKValues` and `topKIndices` GPU implementation. Currently I need to switch between backends for executing classification and computing the top 5 classes in one graph (like in the squeezenet demo).

Would eventually be great to automatically fallback to the CPU implementation, if available.

_Copied from original issue: tensorflow/tfjs-core#488_

**Comment by nsthorat** [comment]
_From @chaosmail on January 13, 2018 16:30_

@dsmilkov do you have any idea how to implement this elegantly on WebGL without reading and sorting the entire texture in a single fragment shader? Would be nice if one could use the depth buffer to automatically sort the values, but I am not sure if this could work also for k > 1.

**Comment by nsthorat** [comment]
_From @dsmilkov on January 15, 2018 1:37_

To be honest, I have no idea how this is done on the GPU. I can chat with a GPU expert at work next week. In the meantime, I'd suggest the standard Google/Stack Overflow search :)

**Comment by nsthorat** [comment]
_From @charlesnicholson on January 15, 2018 4:1_

I'm not aware of any methods that can exploit z-buffering, but that would be nice.

A single fragment shader will work, but can't exploit any GPU parallelism.

The standard approach to max ("topk, k=1") in pre-shared-memory GPU environments is to do a downsampling ping-pong. For a square texture of width and length 2^N, render N passes, giving N draw calls (log2(texture size)). The input to any phase i is the max texture of resolution 2^i with i in [1, N], and the output render target is resolution 2^(i-1).

The pixel shader for each phase simply investigates its four neighbors and stores the max value in one of its own output texture channels. Repeat until the output texture is 1x1, and the final texture's only texel contains the global maximum value. This parallelizes well on the GPU because you get as many cores / wavefronts as you have tiles in your ndarray.

This approach does require 33% more texture memory and a fair amount more bookeeping: you need to create what's effectively a mip-chain from the input ndarray all the way down to the final 1x1 reduction. These mip chains could be re-used across any ndarray with compatible sizes, though, if they stay resident on the GPU. The high-level typescript operation would be a for loop of these ping-pong draw calls.

It's easy-but-slow to extend this approach to top-k: simply run this entire reduction operation k times, store the most recent pass's max 1x1 texture in the pixel shader as a read-only sampler, and reject any texels >= k. It will likely stay resident in texture cache for all cores, since every pixel shader samples it identically and it's only 1x1. You'll need a special final draw call per pass that copies the value of the 1x1 texture into the correct texcoord (k, 0.5) of the final top-k texture, but that seems straightforward.

It's a little more challenging to parallelize these k reductions in the pixel shader, but possible? If the output texture has 4 channels, a single reduction operation could end up with top-4. I think you could do it with a set of cascading if statements? This sounds tricky because each neighbor that you sample would contain the top-4 of its render call, so you'd probably have to write a manually-unrolled 4 x 4-way merge. Fun.

**Comment by nsthorat** [comment]
_From @charlesnicholson on January 15, 2018 4:27_

Top-k indices would be tougher with a top-4 approach. Because you only have 4 channels (tops) in your output texture, you don't have anywhere to put the st texel coordinates that represent the index. You could just do max ("top-1") each time, and use R for max, G for s-coord, and B for t-coord, maybe. That forces you to do k reductions for top-k.

I guess once you've found the k-values for any resolution phase, you could do another draw call with the previous k-values and the new k-values as input samplers, and output to an identically-sized k-indices texture as output. That shader would just figure out which neighbor was chosen, and write that neighbor's st coords to its output rg channels or something.

**Comment by nsthorat** [comment]
_From @dsmilkov on January 15, 2018 14:15_

Thanks Charles!

For top-1 (`math.max/min()`), we actually did it with logN passes, (see [recursive reduce call](https://github.com/PAIR-code/deeplearnjs/blob/master/src/math/backends/backend_webgl.ts#L454)). The only difference is that we did it in the logical space, on a logically flatten array and left the shader sampler figure out the 2D sampling.

For v0, I like your suggestion of running this reduction step K times, serially, just so we have this op on the GPU.

We should definitely do the top-4 parallelization when we add capability to sample from packed textures in our shader compiler/splicer, so we still stay in logical space (from the user perspective).

**Comment by nsthorat** [comment]
_From @charlesnicholson on January 15, 2018 14:22_

Let's grab a whiteboard tomorrow and figure it out :)

**Comment by nsthorat** [comment]
_From @chaosmail on January 15, 2018 17:55_

thanks for the detailed insights, seems there is no 'easy' way to do distributed global sorting :)

just throwing another thought in here, concerning `k=1` and Depth Buffer:
* if we could transform all values of the NDArray to a range of [0, 1]
  (luckily softmax returns exactly this range)
* and for each vertex write to position 0,0 in a 1x1 texture
* and use color channels for value and array index
* and set gl_FragDepth to `1 - value`
* and disable blend mode
* shouldnt we obtain the max value and index at position 1x1 in the resulting framebuffer?

**Comment by nsthorat** [comment]
_From @charlesnicholson on January 15, 2018 19:34_

I think I see what you're getting at, but I'm not sure I understand it fully yet.

If I understand you correctly, you're proposing binding 1x1 color + depth targets to the framebuffer? If that's the case, then you only get one pixel to shade, right?

When you say "for each vertex write to position 0,0 in a 1x1 texture", I'm unsure what you mean.

There are only 6 verts total, no matter how big the ndarray: 2 triangles which form a unit square in xy. We size the viewport exactly to the output ndarray, so that each pixel maps precisely to one output ndarray texel.

The vertex shader is run 4 times, one for each unique vertex. After that, the pixel shader is run once per output ndarray pixel, which I think is the only opportunity to exploit GPU parallelism.

Sorry if I'm being dense and just totally missing the middle part of your idea. I think I get the big picture, though, which is "render the whole ndarray to a single pixel and let the depth buffer naturally accumulate the 'closest' value, which will be the max".

**Comment by nsthorat** [comment]
_From @chaosmail on January 15, 2018 21:27_

Thanks for clarifying the vertex setup!

Yeah, I was thinking of using m x n vertices (points) as input for an m x n NDArray (this might be a bad idea though). I sketched the approach that I had in mind (on the right) vs 1 pass with the unit square (on the left).

![max_via_depth](https://user-images.githubusercontent.com/2969388/34961903-9ca254ce-fa41-11e7-8675-63cf8d0cf2c2.jpeg)

The idea is based on the assumption that the fragment shader is executed for each input vertex using `GL_POINTS`. It's been some time since I studied Computer Graphics, so I am not entirely sure if this is correct.

**Comment by nsthorat** [comment]
_From @charlesnicholson on January 15, 2018 22:3_

Oh, interesting, and clever :) Now I see what you're getting at, thanks for taking the time to sketch it out.

Let me rephrase it so you can tell if I understand correctly:

"If you could project every single normalized array value onto a beam defined by the orthographic camera location and the middle of the framebuffer's only output pixel, the z-buffer hardware would only accept values closer to the near clip plane than the previous closest value. The resulting value is the global min/max."

To do that, you'd have to:

1. Allocate a vertex buffer where each vertex has attributes for the (m, n) ndarray coordinates, as well as the (x) ndarray value at that coordinate.
2. Populate the vertex buffer with this data and upload it to the GPU.
3. Configure an inverse depth buffer, so that largest z-values are closest (for a max top-k anyway). Configure the near clip plane to be 1.0 and the far clip plane to be 0.
4. Load a vertex shader that simply writes the (m, n, x) attributes to the fragment shader inputs.
5. Load a fragment shader that simply write the (m, n, x) attributes to the gl_FragColor output.

OK, that makes sense to me. I think it could probably work, but maybe not efficiently with the current deeplearn.js architecture. I think there are 2 big obstacles that might be show-stoppers:

1. Allocating all of those vertex buffers is going to be a pretty significant increase to the GPU memory footprint! ~8-12 bytes per vertex * 1024 * 1024 (for an [1024,1024] ndarray) is going to clock in at 8-12MB per ndarray!

2. Avoiding a GPU -> CPU -> GPU round trip is going to be impossible, I think! Deeplearn.js is really fast because it doesn't stall out the GPU by reading back all of the pixels from every draw call; it works very hard to keep all intermediate computations resident on the GPU in textures, and only issues readbacks when the chain of computations is complete. If we need to feed ndarray data that's on the GPU into a vertex buffer, that's a GPU => CPU => GPU interlock situation. WebGL 2.0 gets you a little bit closer with the [Transform Feedback](https://webgl2fundamentals.org/webgl/lessons/webgl2-whats-new.html) feature, but unfortunately that only lets vertex shaders write back into their vertex buffers. There isn't a mechanism I'm aware of to write texel data back into vertex buffer data without doing a full CPU readback.

I think all of this kind of stuff is why the ping-pong reduction approach is the best currently-known solution. It's a lot of draw calls, but the data all stays resident on the GPU, and it works with bizarre "my texels are /actually/ array elements" situations like ours :)

**Comment by dsmilkov** [comment]
Fixed in #1172. Folks, feel free to reopen a new issue about speeding up TopK on GPU specifically.
